sympy/concrete/products.py

```diff
@@ -399,6 +399,18 @@ def _eval_product_direct(self, term, limits):
         (k, a, n) = limits
         return Mul(*[term.subs(k, a + i) for i in range(n - a + 1)])
 
+    def _eval_derivative(self, x):
+        if self.is_zero:
+            return S.Zero
+        from sympy.concrete.summations import Sum, Dummy
+        i = Dummy('i')
+        f = self.function
+        limits = list(self.limits)
+        _, a, b = limits[0]
+        if any(x in limit[1:].free_symbols for limit in limits):
+            return None
+        return Sum(f.subs(_, i).diff(x) * Product(f, (_, a, i - 1)) * Product(f, (_, i + 1, b)), (i, a, b), **self.assumptions0)
```

Comment:
This would produce wrong answers if `limits` has more than one element. I thought that `Sum` had a similar problem but it does not. Looking at the code of `Sum`: 
- it pops `limits[0]` from the stack
- checks if there are still limits remaining
    * if there are limits remaining, then it replaces `f` with `Sum(f, remaining_limits)`
- it calls `Derivative(f, ...)` instead of `f.diff(x)`

This way, it would keep all the sums as required. 

I think a similar mechanism should be a adopted here. It might even be better to copy the code and change it as you need to. That way, if there is a bug in `Sum.diff` then there will be a similar bug in `Product.diff` and they can be fixed in a similar way. I think it might help with maintainability.

Reply:
> This would produce wrong answers if `limits` has more than one element.
Can you please provide an example?
I ran it for:
```
In [2]: Product(x*y, (x, 1, z), (y, 1, m)).diff(z).doit()
Out[2]:
    z   m
m⋅m! ⋅z! ⋅Γ(z + 1)⋅polygamma(0, z + 1)             z   m
────────────────────────────────────── + log(m!)⋅m! ⋅z!
                  z!

In [3]: Product(x*y, (x, 1, z), (y, 1, m)).doit().diff(z)
Out[3]:
    z   m
m⋅m! ⋅z! ⋅Γ(z + 1)⋅polygamma(0, z + 1)             z   m
────────────────────────────────────── + log(m!)⋅m! ⋅z!
                  z!
```
This seems to work though.
Thanks.

Reply:
With this, you are differentiating with respect to one of the indexes of the summation. So the `_eval_derivaitve` returns `None` and the value becomes `Derivative(Product(x*y, (x, 1, z), (y, 1, m)), z)` and calling doit will simplify the product and then it takes the derivative. 

I'll quickly come up with an example that shows this issue.

Reply:
```python
from sympy import *
x, y = symbols("x y")
i, j, m, n = symbols("i, j, m, n", integer=True)

expr = Product(sin(x*i/j), (i, 1, m), (j, 1, n))
print(expr.diff(x))
# output:
# Sum(_i*cos(_i*x/j)*Product(sin(i*x/j), (i, 1, _i - 1))*Product(sin(i*x/j), (i, _i + 1, m))/j, (_i, 1, m))
```
As you can see, the `(j, 1, n)` part has completely disappeared from the output.

Reply:
@Maelstrom6 Thank you for correcting, I went ahead as you suggested to start with code in `Sum`.
The implementation below seems to works for the case you gave and for the previous cases:

```
def _eval_derivative(self, x):
        from sympy.concrete.summations import Sum, Dummy, Symbol, Derivative
        if isinstance(x, Symbol) and x not in self.free_symbols:
            return S.Zero
        f, limits = self.function, list(self.limits)
        limit = limits.pop(-1)
        if limits:
            f = self.func(f, *limits)
        _, a, b = limit
        if x in a.free_symbols or x in b.free_symbols:
            return None
        h = Dummy()
        rv = Sum( Product(f, (_, a, h - 1)) * Product(f, (_, h + 1, b)) * Derivative(f, x, evaluate=True).subs(_, h), (h, a, b))
        return rv
```

```
In [1]: i,j=symbols('i j')

In [2]:  expr = Product(sin(x*i/j), (i, 1, n), (j, 1, m))

In [3]: print(expr.diff(x))
Sum(Product(sin(i*x/j), (i, 1, n), (j, 1, _Dummy_23 - 1))*Product(sin(i*x/j), (i, 1, n), (j, _Dummy_23 + 1, m))*Sum(_Dummy_24*cos(_Dummy_24*x/_Dummy_23)*Product(sin(i*x/_Dummy_23), (i, 1, _Dummy_24 - 1))*Product(sin(i*x/_Dummy_23), (i, _Dummy_24 + 1, n))/_Dummy_23, (_Dummy_24, 1, n)), (_Dummy_23, 1, m))
```
Kindly give your opinion.
Thanks.